### PR TITLE
discovery: generate service name from container tags

### PR DESCRIPTION
**Pull request by Yumasi**
### What does this PR do?

This PR makes service discovery generate service names of containerised services using the container tags.

### CODE REVIEW

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 1]
```diff
@@ -613,6 +614,65 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
...
+		{"short_image", nil},
+		{"kube_container_name", nil},
+		{"kube_deployment", nil},
+		{"kube_service", nil},
+	}
+
+	service.ContainerID = id
+	log.Debugf("Found container id for %v: %v", service.Name, id)
+	for _, c := range containers {
+		if c.Id != id {
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:640]: Do you think it would be worth it to create a temporary id-to-container map to avoid looping through the container list for every service?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 2]
```diff
@@ -613,6 +614,65 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
...
+
+				value := tag[sepIndex+1:]
+				tagsPriority[i].tagValue = &value
+				break
+			}
+		}
+
+		for _, tag := range tagsPriority {
+			if tag.tagValue != nil {
+				service.GeneratedName = *tag.tagValue
```
[the remark this answers is not in the document]
Comment (Yumasi) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go:666]: > Can we save the generated name in the info cache so that we don't have to do the above computation for every invocation of the endpoint?

I agree with the idea, but in practice, we can't rely on the `generatedName` field being set, because it always will be first by `getService`.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 3]
```diff
@@ -628,22 +712,26 @@ func (s *discovery) getServices() (*[]model.Service, error) {
...
 	alivePids := make(map[int32]struct{}, len(pids))
-	_, _, pidToCid, err := s.containerProvider.GetContainers(1*time.Minute, nil)
+	containers, _, pidToCid, err := s.containerProvider.GetContainers(1*time.Minute, nil)
 	if err != nil {
 		log.Errorf("could not get containers: %s", err)
 	}
 
+	// Build mapping of Container ID to container object to avoid traversal of
+	// the containers slice for every services.
+	containersMap := make(map[string]*agentPayload.Container, 10)
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:722]: Why not `len(containers)` instead of hard-coded 10?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 4]
```diff
@@ -613,6 +616,87 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
...
+	}
+
+	container, ok := containers[id]
+	if !ok {
+		return
+	}
+
+	var serviceName string
+
+	for _, tag := range container.Tags {
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:654]: Can we add some unit tests for this tag parsing logic to check the edge conditions etc? Shouldn't need mocking etc, just a simple unit test. (I think you'll need to break out the inner logic into its own function.)

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 5]
```diff
@@ -613,6 +616,87 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
...
+	if serviceName != "" {
+		s.mux.Lock()
+		defer s.mux.Unlock()
+
+		serviceInfo, ok := s.cache[int32(service.PID)]
+		if !ok {
+			return
+		}
+
+		service.GeneratedName = serviceName
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:693]: Setting `service` doesn't need the lock, and in fact should be done regardless of the cache check since the `service` should probably be enhanced even if the `serviceInfo` is gone from the cache (due to parallel deletion).

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 6]
```diff
@@ -613,6 +616,87 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
...
+		if tag.tagValue == nil {
+			continue
+		}
+
+		serviceName = *tag.tagValue
+		log.Debugf("Using %v:%v tag for service name", tag.tagName, *tag.tagValue)
+		break
+	}
+
+	if serviceName != "" {
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:684]: If we did find the container tags and searched them once but didn't find anything appropriate (so `serviceName` is empty), we should stop searching them again and again. So maybe you should give `GeneratedNameFromContainerTags` a more appropriate name and set it here regardless of `serviceName` being empty or not.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go [hunk 7]
```diff
@@ -203,7 +204,7 @@ func TestBasic(t *testing.T) {
 	var unexpectedPIDs []int
 	expectedPorts := make(map[int]int)
 
-	var startTCP = func(proto string) {
+	startTCP := func(proto string) {
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux_test.go:207]: Looks like your editor is still messing with unrelated code.

**Review by vitkyrka — APPROVED** [review]
